### PR TITLE
[IOPAE-544] Add useFetch react custom hook

### PR DESCRIPTION
**Pull request by giuseppedipinto**
**Custom useFetch hook for client fetch API**
This hook use `@pagopa/openapi-codegen-ts` autogenerated client, therefore makes available all and only APIs specified in the project root opeanapi.yaml _(by operationId)_.

_Parameters:_
- `operationId` openapi operationId
- `requestParams` request parameters _(as specified in openapi)_
- `responseCodec` io-ts expected response codec

_Returns:_
- `data` fetch result of _responseCodec_ type
- `error` in case of request error, decode failure or exceptions
- `loading` as boolean value for request pending state

**Usage example:**
```
const { data, error, loading } = useFetch(
  "regenerateServiceKey",
  {
    serviceId: "aServiceId",
    keyType: SubscriptionKeyTypeEnum.primary
  },
  SubscriptionKeys
);
```
_Error Typing:_
```
export type UseFetchError = {
  kind: "httpError" | "validationError" | "exceptionError";
  status?: number; // is the http response status code
  message: string;
};
```

**Important Notes**
The client is already set up to send the **bearer token** _(currently there is a placeholder here)_ and in case of **401 Unauthorized** it manages the redirect to a dedicated route _(currently there is a placeholder here)_.

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: apps/backoffice/src/hooks/use-fetch.ts [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+      status
+    });
+
+  const fetchData = async (requestParams: any) => {
+    try {
+      const result = await clientWithBearerToken[operationId](requestParams);
+
+      if (result._tag === "Right") {
+        // Check 401 Unauthorized
+        if (result.right.status === 401) {
```
Comment (gunzip) [inline apps/backoffice/src/hooks/use-fetch.ts:124]: I suggest to assign a variable with a ameaningful name to `result.right`

File: apps/backoffice/src/hooks/use-fetch.ts [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+        /** Set a response class type based on http response status code */
+        const httpResponseClassType = manageHttpResponseStatusCode(
+          result.right.status
+        );
+
+        switch (httpResponseClassType) {
+          case "successful":
+            // 2XX successful response status code
+            const decodedResult = responseCodec.decode(result.right.value);
+            decodedResult._tag === "Right"
```
Comment (gunzip) [inline apps/backoffice/src/hooks/use-fetch.ts:138]: fold / bimap?

File: apps/backoffice/src/hooks/use-fetch.ts (hunk 1 above)
Comment (giuseppedipinto) [reply apps/backoffice/src/hooks/use-fetch.ts]: 8376da1

File: apps/backoffice/src/hooks/use-fetch.ts (hunk 2 above)
Comment (giuseppedipinto) [reply apps/backoffice/src/hooks/use-fetch.ts]: 7240f27

File: apps/backoffice/src/hooks/use-fetch.ts [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+            );
+            break;
+          case "clientError":
+          case "serverError":
+            setUseFetchError(
+              "httpError",
+              httpResponseClassType,
+              response.status
+            );
+          default:
```
Comment (rodoviti) [inline apps/backoffice/src/hooks/use-fetch.ts:162]: **suggestion (non-blocking):** remove "useless" default.

If a default block does not have an implementation, simply omit it.
```suggestion
```

**Review by rodoviti — APPROVED** [review]
LGTM

File: apps/backoffice/src/hooks/use-fetch.ts (hunk 3 above)
Comment (giuseppedipinto) [reply apps/backoffice/src/hooks/use-fetch.ts]: 925cfa7

File: apps/backoffice/src/hooks/use-fetch.ts [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+      kind,
+      message,
+      status
+    });
+
+  const fetchData = async (requestParams: any) => {
+    try {
+      const result = await clientWithBearerToken[operationId](requestParams);
+
+      if (result._tag === "Right") {
```
[the remark this answers is not in the document]
Comment (giuseppedipinto) [reply apps/backoffice/src/hooks/use-fetch.ts]: 07fbc1c
_(as we have seen and as the autogenerated client response is defined, it is easier to check the left branch first)_

**Pull request merged**